src/renderer/base/renderer.cpp

```diff
@@ -831,9 +842,19 @@ void Renderer::_PaintBufferOutputHelper(_In_ IRenderEngine* const pEngine,
                 const auto thisPointPatterns = _pData->GetPatternId(thisPoint);
                 const auto thisUsingSoftFont = s_IsSoftFontChar(it->Chars(), _firstSoftFontChar, _lastSoftFontChar);
                 const auto changedPatternOrFont = patternIds != thisPointPatterns || usingSoftFont != thisUsingSoftFont;
-                if (color != it->TextAttr() || changedPatternOrFont)
+                auto origAttr = it->TextAttr();
+                for (auto highlight : highlights)
+                {
+                    if (thisPoint.y == highlight.top && thisPoint.x < highlight.right && thisPoint.x >= highlight.left)
+                    {
+                        origAttr.SetBackground(0x003296ff);
+                        origAttr.SetForeground(0x000000);
+                    }
+                }

```

Comment:
Going through the highlights for each line of text seems fairly expensive. When I wrote this:
> Should we perhaps do the same? Get rid of` _drawSearchSelections` and just manipulate the background/foreground colors of the highlighted text (see second half of `AtlasEngine::PaintBufferLine`)?

I meant that we could simply change `AtlasEngine::PaintSelections` to look like this (untested pseudocode):
```cpp
[[nodiscard]] HRESULT AtlasEngine::PaintSelections(const std::vector<til::rect>& rects) noexcept
try
{
    // Unfortunately there's no step after Renderer::_PaintBufferOutput that
    // would inform us that it's done with the last AtlasEngine::PaintBufferLine.
    // As such we got to call _flushBufferLine() here just to be sure.
    _flushBufferLine();

    for (const auto& rect : rects)
    {
        const auto y = gsl::narrow_cast<u16>(clamp<til::CoordType>(rect.top, 0, _p.s->viewportCellCount.y));
        const auto from = gsl::narrow_cast<u16>(clamp<til::CoordType>(rect.left, 0, _p.s->viewportCellCount.x - 1));
        const auto to = gsl::narrow_cast<u16>(clamp<til::CoordType>(rect.right, from, _p.s->viewportCellCount.x));

        if (rect.bottom <= 0 || rect.top >= _p.s->viewportCellCount.y)
        {
            continue;
        }

        const auto bg = &_p.backgroundBitmap[_p.colorBitmapRowStride * y];
        const auto fg = &_p.foregroundBitmap[_p.colorBitmapRowStride * y];
        std::fill(bg + from, bg + to, 0xff3296ff);
        std::fill(fg + from, fg + to, 0xff000000);
    }
    return S_OK;
}
CATCH_RETURN()
```

and remove all other modifications to `AtlasEngine`. The benefit of this approach is that it directly "decompresses" your selections into the background/foreground color bitmaps.

What do you think about this?

Reply:
I'm not sure how familiar with git, but if you want to revert any parts of your code to the state it was before, you can use `git restore` with the `-s` (source) parameter. For instance, this would undo all recent renderer.cpp changes:
```
git restore -s 328f809 src/renderer/base
```

(Just in case you do think my proposal above is a good idea, I thought this might be helpful.)

Reply:
Just gave it a try,  I like this better than what I had in my last commit.

Reply:
@lhecker I noticed a issue where the highlights did not show until I typed the second character of a search.

It looked like the issue was that this check caused the bitmap image to not be uploaded.
https://github.com/microsoft/terminal/blob/d14524cd4cc4970bb1b6456f9667e2dd661b9854/src/renderer/atlas/BackendD3D.cpp#L966-L969

I guess the first character wasn't passing because the bitmap isn't updated yet when this check happens.
https://github.com/microsoft/terminal/blob/d14524cd4cc4970bb1b6456f9667e2dd661b9854/src/renderer/atlas/AtlasEngine.cpp#L360-L365

If I update my logic to bump the generation,  it looks like it fixes the issue and I couldn't think of any cases where it would cause an extra upload.  I wasn't sure if this would cause the code/optimization to become more brittle though.

```
[[nodiscard]] HRESULT AtlasEngine::PaintSelections(const std::vector<til::rect>& rects) noexcept
try
{
    // Unfortunately there's no step after Renderer::_PaintBufferOutput that
    // would inform us that it's done with the last AtlasEngine::PaintBufferLine.
    // As such we got to call _flushBufferLine() here just to be sure.
    _flushBufferLine();

    for (const auto& rect : rects)
    {
        const auto y = gsl::narrow_cast<u16>(clamp<til::CoordType>(rect.top, 0, _p.s->viewportCellCount.y));
        const auto from = gsl::narrow_cast<u16>(clamp<til::CoordType>(rect.left, 0, _p.s->viewportCellCount.x - 1));
        const auto to = gsl::narrow_cast<u16>(clamp<til::CoordType>(rect.right, from, _p.s->viewportCellCount.x));

        if (rect.bottom <= 0 || rect.top >= _p.s->viewportCellCount.y)
        {
            continue;
        }

        const auto bg = &_p.backgroundBitmap[_p.colorBitmapRowStride * y];
        const auto fg = &_p.foregroundBitmap[_p.colorBitmapRowStride * y];
        std::fill(bg + from, bg + to, 0xff3296ff);
        std::fill(fg + from, fg + to, 0xff000000);

        for (int i = 0; i < 2; i++)
        {
            _p.colorbitmapgenerations[i].bump();
        }
    }
    return S_OK;
}
CATCH_RETURN()
```
Do you guys have a preference on how to deal with this?

I had also looked at moving the _PaintSelections logic before _PaintBufferOutput,  the caused the generation to get bumped but the highlight color also got overwritten here.
https://github.com/microsoft/terminal/blob/d14524cd4cc4970bb1b6456f9667e2dd661b9854/src/renderer/atlas/AtlasEngine.cpp#L363

Reply:
Oh, right, I forgot about that. I'm glad you already figured out that you need to call `_p.colorBitmapGenerations[i].bump();`. 🙂
The performance impact of invalidating too often is minimal so don't worry about it. The `colorBitmapGenerations` thing exists, so we avoid uploading the bitmaps to the GPU when they haven't changed, which reduces power draw on lower end hardware like laptops. But it's not _that_ much.

Your suggested code is fine. You can however "hoist" the `colorBitmapGenerations` loop outside of the outer loop. This however requires us to check for `rects.empty()`, since we don't want to bump the counters when there's no selections that need to be drawn. I'd do it like this:
```cpp
[[nodiscard]] HRESULT AtlasEngine::PaintSelections(const std::vector<til::rect>& rects) noexcept
try
{
    if (rects.empty())
    {
        return S_OK;
    }

    for (const auto& rect : rects)
    {
        // ... your original code ...
    }

    for (int i = 0; i < 2; ++i)
    {
        _p.colorBitmapGenerations[i].bump();
    }
}
CATCH_RETURN()
```

Reply:
thanks!  just updated.  I have been meaning to ask.  Do I need that call to _flushBufferLine?

Reply:
Actually no, but I didn't want to comment on it as it doesn't matter. 🙂

Reply:
got it,  just removed it.  I'll get a chance to squash all of these commits if it gets merged?

Reply:
Yep! We always squash PRs by default.